rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs

```diff
@@ -0,0 +1,1089 @@
+        db: Arc<dyn HyperlaneSequenceIndexerStore<T>>,
+        next_sequence: u32,
+        start_block: u32,
+        index_mode: IndexMode,
+    ) -> Self {
+        // If the next sequence is 0, we're starting from the beginning and haven't
+        // indexed anything yet.
+        let last_indexed_snapshot = LastIndexedSnapshot {
+            sequence: (next_sequence > 0).then(|| next_sequence.saturating_sub(1)),
+            at_block: start_block,
+        };
+
+        Self {
+            chunk_size,
+            latest_sequence_querier,
+            db,
+            last_indexed_snapshot,
+            current_indexing_snapshot: TargetSnapshot {
+                sequence: next_sequence,
+                at_block: start_block,
+            },
+            target_snapshot: None,
+            index_mode,
+        }
+    }
+
+    /// Gets the next range of logs to index.
+    /// If there are no logs to index, returns `None`.
+    /// If there are logs to index, returns the range of logs, either by sequence or block number
+    /// depending on the mode.
+    pub async fn get_next_range(&mut self) -> Result<Option<RangeInclusive<u32>>> {
+        // Fast forward the cursor if necessary.
+        self.fast_forward().await?;
+
+        let (Some(onchain_sequence_count), tip) = self
+            .latest_sequence_querier
+            .latest_sequence_count_and_tip()
+            .await?
+        else {
+            return Ok(None);
+        };
+
+        let current_sequence = self.current_indexing_snapshot.sequence;
+        let range = match current_sequence.cmp(&onchain_sequence_count) {
+            Ordering::Equal => {
+                // We are synced up to the latest sequence so we don't need to index anything.
+
+                // We can update the current indexing snapshot to the tip.
+                // This will let us only index blocks that are likely to have new logs once
+                // there's a new sequence to search for.
+                self.current_indexing_snapshot.at_block = tip;
+
+                None
+            }
+            Ordering::Less => {
+                // The cursor is behind the onchain sequence count, so we need to index.
+
+                // Minus one because this is the sequence we're targeting, not the count.
+                let target_sequence = onchain_sequence_count.saturating_sub(1);
+
+                // Set the target to the highest sequence and tip.
+                // We don't necessarily expect to hit this target in the next query (because we
+                // have limits to the range size based off the chunk size), but we will use it
+                // as an eventual target.
+                self.target_snapshot = Some(TargetSnapshot {
+                    sequence: target_sequence,
+                    at_block: tip,
+                });
+
+                match &self.index_mode {
+                    IndexMode::Block => {
```

Comment:
At this point the indentation level is a bit high for a single function, making it hard to reason about. We should move this into another function. The u32::min range logic can be shared between `Block` and `Sequence`, which would make it even easier to understand

Reply:
will change the indentation but the values in the u32::min range are different, maybe i'm misinterpreting what you mean though

Reply:
> values in the u32::min range are different

Fair, I didn't look closely enough